python-avd/pyavd/_eos_designs/structured_config/overlay/router_bgp.py

```diff
@@ -24,457 +26,360 @@ class RouterBgpMixin(Protocol):
         else:  # other roles are transit / edge
-            address_family_link_state["path_selection"] = {"roles": {"producer": True}}
+            self.structured_config.router_bgp.address_family_link_state.path_selection.roles.producer = True
 
         if self._is_wan_server_with_peers:
-            address_family_link_state["peer_groups"].append({"name": self.inputs.bgp_peer_groups.wan_rr_overlay_peers.name, "activate": True})
-
-        return address_family_link_state
+            peer_groups.append_new(name=self.inputs.bgp_peer_groups.wan_rr_overlay_peers.name, activate=True)
 
-    def _address_family_path_selection(self: AvdStructuredConfigOverlayProtocol) -> dict | None:
+    def _set_address_family_path_selection(self: AvdStructuredConfigOverlayProtocol) -> None:
         if not self.shared_utils.is_wan_router:
-            return None
-
-        address_family_path_selection = {
-            "peer_groups": [
-                {
-                    "name": self.inputs.bgp_peer_groups.wan_overlay_peers.name,
-                    "activate": True,
-                },
-            ],
-            "bgp": {"additional_paths": {"receive": True, "send": "any"}},
-        }
+            return
 
-        if self._is_wan_server_with_peers:
-            address_family_path_selection["peer_groups"].append({"name": self.inputs.bgp_peer_groups.wan_rr_overlay_peers.name, "activate": True})
+        peer_groups = self.structured_config.router_bgp.address_family_path_selection.peer_groups
+        peer_groups.append_new(name=self.inputs.bgp_peer_groups.wan_overlay_peers.name, activate=True)
 
-        return address_family_path_selection
+        self.structured_config.router_bgp.address_family_path_selection.bgp.additional_paths._update(receive=True, send="any")
 
-    def _address_family_rtc(self: AvdStructuredConfigOverlayProtocol) -> dict | None:
+        if self._is_wan_server_with_peers:
+            peer_groups.append_new(name=self.inputs.bgp_peer_groups.wan_rr_overlay_peers.name, activate=True)
+
+    def _set_address_family_rtc(self: AvdStructuredConfigOverlayProtocol) -> None:
         """
         Activate EVPN OVERLAY peer group and EVPN OVERLAY CORE peer group (if present) in address_family_rtc.
 
         if the evpn_role is server, enable default_route_target only
         """
         if not self.inputs.evpn_overlay_bgp_rtc:
-            return None
-
-        address_family_rtc = {}
-
-        peer_groups = []
-        evpn_overlay_peers = {"name": self.inputs.bgp_peer_groups.evpn_overlay_peers.name}
+            return
+        peer_groups = self.structured_config.router_bgp.address_family_rtc.peer_groups
+        peer_groups.append_new(name=self.inputs.bgp_peer_groups.evpn_overlay_peers.name)
         if self.shared_utils.overlay_evpn_vxlan is True:
-            evpn_overlay_peers["activate"] = True
+            peer_groups.obtain(self.inputs.bgp_peer_groups.evpn_overlay_peers.name).activate = True
```

Comment:
We should not use obtain here. Instead move the if above the append_new and set activate. But it seems weird that we would not always activate the peer-group if we add it. So IMO we should only add the peer-group here if we also intend to activate it.

Please also avoid the obtain below. Maybe just create new peer-group items and append them after you have set the attributes you need.

Reply:
Fixed!

Reply:
Still need to remove obtain from the other address families. I like the logic of evpn now, so use the same pattern.